### DISCUSSION TITLE
[gatsby-theme-minimal-blog] shadowing additional frontmatter

### DISCUSSION THREAD

**Discussion by myedibleenso** [LekoArts/gatsby-themes#829] [Questions]
@LekoArts [shared a concise solution for adding frontmatter to a Post](https://github.com/LekoArts/gatsby-themes/issues/438#issuecomment-653674876).  Following that approach, I was able to add a new frontmatter field that is showing up in interactive queries (see my modified `gatsby-node.js` below). 

```javascript
// My gatsby-node.js
const addFrontmatterField = (name) => async (source, args, context) => {
  const mdxNode = context.nodeModel.getNodeById({
    id: source.parent,
  })
  return mdxNode.frontmatter[name]
}

exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes, createFieldExtension } = actions

  createFieldExtension({
    name: `addFrontmatterField`,
    args: {
      name: `String!`,
    },
    extend({ name }) {
      return {
        resolve: addFrontmatterField(name),
      }
    },
  })

  createTypes(`
    interface Post @nodeInterface {
      author: String
    }
    
    type MdxPost implements Node & Post {
      author: String @addFrontmatterField(name: "author")
    }
  `)
}
```

Unfortunately, I'm still unable to access the new field when shadowing.  

Assuming one wanted to access this new `author` field in `blog-list-item.tsx`, for example, what changes would be necessary?  I tried the following unsuccessfully:

```jsx
/** @jsx jsx */
import * as React from "react"
import { jsx, Box } from "theme-ui"
import { Link, graphql } from "gatsby"
import ItemTags from "@lekoarts/gatsby-theme-minimal-blog/src/components/item-tags";

type BlogListItemProps = {
  post: {
    slug: string
    title: string
    author?: string
    date: string
    excerpt: string
    description: string
    timeToRead?: number
    tags?: {
      name: string
      slug: string
    }[]
  }
  showTags?: boolean
}

const BlogListItem = ({ post, showTags = true }: BlogListItemProps) => (
  <Box mb={4}>
    <Link to={post.slug} sx={(t) => ({ ...t.styles?.a, fontSize: [1, 2, 3], color: `text` })}>
      {post.title}
      <p> Author: {post.author ? post.author : `????` } </p>
    </Link>
    <p sx={{ color: `secondary`, mt: 1, a: { color: `secondary` }, fontSize: [1, 1, 2] }}>
      <time>{post.date}</time>
      {post.tags && showTags && (
        <React.Fragment>
          {` — `}
          <ItemTags tags={post.tags} />
        </React.Fragment>
      )}
    </p>
  </Box>
)

export default BlogListItem
```

I thought perhaps I needed to also shadow `@lekoarts/gatsby-theme-minimal-blog-core/src/templates/blog-query.tsx`, but that didn't seem to work either.  In fact, that gives me a warning: 

```
The GraphQL query in the non-page component "/app/src/@lekoarts/gatsby-theme-minimal-blog-core/src/templates/blog-query.tsx" will not be run.
```

**Comment by md-weber** [comment]
Exactly what I needed; thanks to you, I found how to add the `author` field to my posts, and if I try to shadow the blog-query file, it throws me errors that the query can only be used on a page level.

I tried to edit the parent file and was able to receive the author. Currently, I am trying to work with StaticQuery but was not yet successful. I will try tomorrow a bit more. I will keep you posted if I find a good solution.

**Comment by half0wl** [comment]
I figured out how to do this via https://github.com/LekoArts/gatsby-themes/issues/438#issuecomment-653674876. There's no modification required to gatsby-node at all - you can just inject a field into the frontmatter and query for it via GQL. Here's what I did:

Shadow the homepage query to include mdx frontmatter (`@lekoarts/gatsby-theme-minimal-blog-core/templates/homepage-query.tsx`):
```gql
<truncated>
      nodes {
        <truncated>
        description
        tags {
          name
          slug
        }
+        ... on MdxPost {
+          parent {
+            ... on Mdx {
+              frontmatter {
+                $YOUR_FIELD
+              }
+            }
+          }
+        }
      }
<truncated>
```

This makes it available in the `posts` object downstream in this shape:

```typescript
export type MBHomepageProps = {
  posts: {
    slug: string
    title: string
    date: string
    excerpt: string
    description: string
    timeToRead?: number
    tags?: {
      name: string
      slug: string
    }[]
+    parent: {
+      frontmatter: {
+        [key: string]: any
+      }
+    }
  }[]
}
```

You can then access your custom field via `post.parent.frontmatter.$YOUR_FIELD`.

I used this to add a `featured` field to the frontmatter for pinning blog posts in my homepage: https://github.com/half0wl/raychen.io/commit/559567fad7af5882f6e77194a929c7380b6a7cb5.

This is an old discussion, but I hope it helps someone stumbling here by Google :-)
